File: 1475. Final Prices With a Special Discount in a Shop.cpp
class Solution {
public:
    vector<int> finalPrices(vector<int>& prices) 
    {
        int n = prices.size();
        vector<int> stact, ans(n);
        for(int i = 0; i < n; i++)
        {
            while(!stact.empty() and prices[stact.back()] >= prices[i])
            {
                ans[stact.back()] = prices[stact.back()] - prices[i];
                stact.pop_back();
            }
            stact.push_back(i);
        }

        for(auto &i : stact)
            ans[i] = prices[i];

        return ans;
    }
};

File: 1475. Final Prices With a Special Discount in a Shop.py
class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        n = len(prices)
        stact, ans = [], [0] * n

        for i in range(n):
            while stact and prices[stact[-1]] >= prices[i]:
                ans[stact[-1]] = prices[stact[-1]] - prices[i]
                stact.pop()
            stact.append(i)
            
        for idx in stact:
            ans[idx] = prices[idx]
        
        return ans